File: Collections/QuickUnion.cs
using System;

namespace Collections
{
    /// <summary>
    /// Data structure that uses the weighted quick union algorithm to solve the Dynamic Connectivity problem.
    /// </summary>
    public class QuickUnion<N> : IUnionFind<N>
    {
        private readonly HashMap<N, int> rank = new HashMap<N, int>(); // rank by size; rank[i] is the size of the tree rooted at i
        private readonly HashMap<N, N> id = new HashMap<N, N>(); // id[i] is the parent of i in their connected cpt. tree

        public int Size() => id.Size();

        /// <summary>
        /// Create a <code>QuickUnion</code> object for the given set of (hashable) nodes.
        /// </summary>
        public QuickUnion(params N[] nodes)
        {
            // Initially there are n trees (singleton components) in the forest
            foreach (var n in nodes)
            {
                id[n] = n;
                rank[n] = 1;
            }
        }

        private N Root(N k)
        {
            while (!k.Equals(id[k]))
            {
                id[k] = id[id[k]];  // Partial path compression
                k = id[k];
            }  // A root node is its own parent
            return k;
        }

        public bool Connected(N i, N j)
        {
            if (i == null || j == null)
                throw new ArgumentNullException();

            if (!id.Contains(i) || !id.Contains(j))
                throw new ArgumentException();

            return Root(i).Equals(Root(j));
        }

        public void Union(N i, N j)
        {
            if (i == null || j == null)
                throw new ArgumentNullException();

            if (!id.Contains(i) || !id.Contains(j))
                throw new ArgumentException();

            var rootI = Root(i);
            var rootJ = Root(j);
            if (rootI.Equals(rootJ)) return;  // i and j are already connected

            if (rank[rootI] > rank[rootJ])
            {
                rank[rootI] += rank[rootJ];
                id[rootJ] = rootI;
            }
            else
            {
                rank[rootJ] += rank[rootI];
                id[rootI] = rootJ;
            }  // Prevent tall trees by connecting the smallest cpt. to the largest
        }
    }
}

File: src/main/java/in/bfdes/collections/QuickUnion.java
package in.bfdes.collections;

import java.util.HashMap;
import java.util.Map;

public class QuickUnion<T> implements UnionFind<T> {
    private final Map<T, Integer> ranks = new HashMap<>();
    private final Map<T, T> parents = new HashMap<>();

    public QuickUnion(T[] elements) {
        for (T e : elements) {
            parents.put(e, e);
            ranks.put(e, 1);
        }
    }

    private T root(T e) {
        while (!e.equals(parents.get(e))) {
            e = parents.get(e);
        }
        return e;
    }

    @Override
    public boolean connected(T i, T j) {
        if (i == null || j == null)
            throw new IllegalArgumentException();

        if (!parents.containsKey(i) || !parents.containsKey(j))
            throw new IllegalArgumentException();

        return root(i) == root(j);
    }

    @Override
    public void union(T i, T j) {
        if (i == null || j == null)
            throw new IllegalArgumentException();

        if (!parents.containsKey(i) || !parents.containsKey(j))
            throw new IllegalArgumentException();

        T rootI = root(i);
        T rootJ = root(j);
        if (rootI == rootJ) return;  // `i` and `j` are already connected
        int rankI = ranks.get(rootI);
        int rankJ = ranks.get(rootJ);

        if (rankI > rankJ) {
            ranks.put(rootI, rankI + rankJ);
            parents.put(rootI, rootJ);
        } else {
            ranks.put(rootJ, rankI + rankJ);
            parents.put(rootJ, rootI);
        }

        parents.put(rootI, rootJ);
    }
}
